### PR TITLE
Findbugs fixes

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: -
- Related Pull Requests: #137 

* * *

## What
This Pull Request adds to the repository a fully enabled FindBugs configuration

## Why
This Pull Request is needed because the old configuration was almost entirely disabled and the tool shows you potential bugs

## How
This feature can be viewed/tested within the project by building the project. Findbugs run on the bytecode instead of the source code, so just running check or even findbugs won't work

## Alternative implementation
Other implementations that I've have considered are -

## Notes

The serialization proxy pattern is only used for objects without circular dependencies, because it clearly doesn't work otherwise. For the other serialize objects we use readObject and writeObject methods for improved efficiency and security.
In this PR stas' fix-bluetooth PR is merged because it has a lot to do with the Bluetooth class (of course) and that class had to be completely refractored because of the whole goal of this pr: fix findbugs.

Old notes:
The new SerializationProxy is used to greatly reduce the likelikehood of bugs and security problems introduced by implementing the Serialization interface. Without going too much in depth here, have a look at item 74 for an explanation about the problems with Serialization and at item 78 for the serialization proxy pattern: [Effective Java](http://index-of.es/JAVA/Effective%20Java.pdf). Effective Java is widely regarded as the best book written about the Java programming language ever, written by one of the main developers of Java (Joshua Bloch).
The serialization proxy is serialized instead of the object itself and largely eliminates the extralinguistic character of serialization, because the deserialized instance is created using the same constructors as any othe rinstance. That frees us from having to separately ensure that deserialized instances obey the class's invariants.

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java [hunk 1]
```diff
@@ -48,36 +52,12 @@ public RSAKey(final Map<String, String> map) throws IllegalArgumentException {
...
-     * @throws Exception If the provided {@link IKey} is not a {@link RSAKey}.
-     */
-    public static String getExponent(final IKey key) throws IllegalArgumentException {
-        final String type = key.getType();
-        if (!type.equals(ConstantKeywords.RSA_KEY)) {
-            throw new IllegalArgumentException();
-        }
-
-        return ((RSAKey) key).exponent;
+    public String getExponent() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java]: I think this is a very bad idea since we want to work with the `IKey` interface as much as possible. The static methods allows you to more easily get the exponent/modules once you know the `IKey` is a `RSAKey`, so you don't have to manually cast etc.

File: app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java [hunk 2]
```diff
@@ -6,22 +6,29 @@
...
+    private final IServiceLocator serviceLocator;
+    private transient final String androidFilesDir;
 
     /**
      * Prevents construction from outside the class.
      *
      * @param serviceLocator Can be used to get access to other modules
      */
     private Constants(final IServiceLocator serviceLocator) {
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java]: The `serviceLocator` on `this` is never used, right?

File: app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java [hunk 3]
```diff
@@ -53,4 +60,37 @@ public String getDatabaseUserdataPath() {
...
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     * Used for the Serialization Proxy Pattern.
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java]: What is different about this serializable proxy that it needs this suppression (other serializable proxies don't have this suppression, e.g. `SimpleKey` of `RSAKey`)?

File: app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java [hunk 4]
```diff
@@ -21,6 +27,7 @@
     // This servicelocator will be used later on probably
     private EncryptorAdapter(final IServiceLocator serviceLocator) {
         LOGGER.info("Initialized");
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java]: The log should come after this assignment :wink:

File: app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/KeyGeneratorAdapter.java [hunk 5]
```diff
@@ -30,6 +35,7 @@
     // This servicelocator will be used later on probably
     private KeyGeneratorAdapter(final IServiceLocator serviceLocator) {
         LOGGER.info("Initialized");
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/KeyGeneratorAdapter.java]: The log should come after this assignment :wink:

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 6]
```diff
@@ -48,13 +51,15 @@
...
     private final String profilesPath;
+    private final IServiceLocator serviceLocator;
 
     /**
      * Prevents construction from outside the class.
      *
      * @param serviceLocator Can be used to get access to other modules
      */
     private GsonDatabaseAdapter(final IServiceLocator serviceLocator) {
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: The `serviceLocator` on `this` is never used, right? Also the log should come after this assignment :wink:

File: app/src/main/java/com/nervousfish/nervousfish/modules/filesystem/AndroidFileSystemAdapter.java [hunk 7]
```diff
@@ -21,6 +27,7 @@
     // This servicelocator will be used later on probably
     private AndroidFileSystemAdapter(final IServiceLocator serviceLocator) {
         LOGGER.info("Initialized");
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/filesystem/AndroidFileSystemAdapter.java]: The log should come after this assignment :wink:

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyBluetoothHandler.java [hunk 8]
```diff
@@ -22,6 +28,7 @@
     private DummyBluetoothHandler(final IServiceLocator serviceLocator) {
         super();
         LOGGER.info("Initialized");
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyBluetoothHandler.java]: The log should come after this assignment :wink:

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyNFCHandler.java [hunk 9]
```diff
@@ -22,6 +28,7 @@
     private DummyNFCHandler(final IServiceLocator serviceLocator) {
         super();
         LOGGER.info("Initialized");
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyNFCHandler.java]: The log should come after this assignment :wink:

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java [hunk 10]
```diff
@@ -22,6 +28,7 @@
     private DummyQRHandler(final IServiceLocator serviceLocator) {
         super();
         LOGGER.info("Initialized");
+        this.serviceLocator = serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java]: The log should come after this assignment :wink:

File: config/quality.gradle [hunk 11]
```diff
@@ -36,24 +36,20 @@ checkstyle {
...
     reports {
         xml.enabled = false
 		xml.withMessages = true
-        html.enabled = true
-        html {
-            destination "$reportsDir/findbugs/findbugs.html"
-        }
+        html.enabled = !xml.isEnabled()
+        xml.destination "$reportsDir/findbugs/findbugs.xml"
+        html.destination "$reportsDir/findbugs/findbugs.html"
```
Comment (ericcornelissen) [inline config/quality.gradle]: How about:
```gradle
reports {
  xml {
    enabled = true
    withMessages = true
    destination "$reportsDir/findbugs/findbugs.xml"
  }
  html {
    enabled = true
    destination "$reportsDir/findbugs/findbugs.html"
  }
}
```

File: config/quality/findbugs/findbugs-filter.xml [hunk 12]
```diff
@@ -8,13 +8,11 @@
...
-    <!-- All bugs in test classes, except for JUnit-specific bugs -->
     <Match>
         <Class name="~.*\.*Test" />
-        <Not>
-            <Bug code="IJU" />
-        </Not>
     </Match>
-
-
+    <!-- TODO Should be removed after #137 is merged -->
```
Comment (TheBonheurs) [inline config/quality/findbugs/findbugs-filter.xml]: Don't we have classes taht should be excluded from this check? Or mayeb in the future?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java [hunk 13]
```diff
@@ -34,4 +41,39 @@ private DummyQRHandler(final IServiceLocator serviceLocator) {
...
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     * Used for the Serialization Proxy Pattern.
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java]: comment on the suppress?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java [hunk 14]
```diff
@@ -34,4 +41,39 @@ private DummyQRHandler(final IServiceLocator serviceLocator) {
...
+    /**
+     * Ensure that no instance of this class is created because it was present in the stream. A correct
+     * stream should only contain instances of the proxy.
+     */
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java]: this class could be {@link this} or something like that, see the same below

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyBluetoothHandler.java [hunk 15]
```diff
@@ -34,4 +41,39 @@ private DummyBluetoothHandler(final IServiceLocator serviceLocator) {
...
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     * Used for the Serialization Proxy Pattern.
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyBluetoothHandler.java]: Comment suppression

File: app/src/main/java/com/nervousfish/nervousfish/modules/filesystem/AndroidFileSystemAdapter.java [hunk 16]
```diff
@@ -33,4 +40,39 @@ private AndroidFileSystemAdapter(final IServiceLocator serviceLocator) {
...
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     * Used for the Serialization Proxy Pattern.
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/filesystem/AndroidFileSystemAdapter.java]: Comment suppression

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 17]
```diff
@@ -300,4 +305,38 @@ private void initializeUserdata() throws IOException {
...
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     * Used for the Serialization Proxy Pattern.
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Comment suppression

File: app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/KeyGeneratorAdapter.java [hunk 18]
```diff
@@ -64,4 +68,38 @@ public static KeyPair generateRSAKeyPair(final String name) throws NoSuchAlgorit
...
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     * Used for the Serialization Proxy Pattern.
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/KeyGeneratorAdapter.java]: Comment Suppression

File: app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java [hunk 19]
```diff
@@ -33,4 +40,38 @@ private EncryptorAdapter(final IServiceLocator serviceLocator) {
...
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     * Used for the Serialization Proxy Pattern.
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java]: Comment suppression

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 20]
```diff
@@ -137,47 +122,46 @@ private void fillDatabaseWithDemoData() throws IOException {
...
      */
-    @NonNull
-    @Override
-    public View getView(final int position, final View convertView, @NonNull final ViewGroup parent) {
-        View v = convertView;
-
-        if (v == null) {
-            final LayoutInflater vi = LayoutInflater.from(getContext());
-            v = vi.inflate(R.layout.contact_list_entry, null);
+    private static final class ContactListAdapter extends ArrayAdapter<Contact> {
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Should this be changed in this pr?

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java [hunk 21]
```diff
@@ -71,4 +75,37 @@ public int hashCode() {
...
+     */
+    private void readObject(final ObjectInputStream stream) throws InvalidObjectException {
+        throw new InvalidObjectException("Proxy required.");
+    }
+
+    /**
+     * Represents the logical state of this class and copies the data from that class without
+     * any consistency checking or defensive copying.
+     */
+    private static final class SerializationProxy implements Serializable {
```
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Of course, have a look at the updated PR description

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java (hunk 1 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java]: I'm not so sure. When you get the exponent / modulus from a key and you work the way as you proposed, then you have no idea if the exponent / modulus you got are valid. You'll have to write code around your calls to the getters to check if they are valid (so from an RSA key) or invalid (from, for example, a SimpleKey). That's very error prone and a design smell.

I would very much like to propose a better way to solve this, maybe with a Visitor pattern, but because the getExponent() and getModulus() methods are not used anywhere I cannot. I think it's best to delete them for now, because they are not used, and when they are needed in the future, that we'll then find an elegant solution that suits that situation.

File: app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java (hunk 2 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java]: That's right. However, the reference is just 8 bytes so practically nothing and just keeping the reference keeps the code consistent with the other modules and makes it easier in the future to use the service locator

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java]: Actually, just thought a bit more about why I did this, as consistency is no very good reason. The real reason a bit more serious.
When Constants is serialized we're essentially making a copy of Constants and when we're deserializing it we're essentially turning that copy in a new instance. This can be done using a very unsafe way, using a extralinguistic mechanism, namely the ordinary serialization. However, a much safer way is to do it by using the proxy serialization pattern. This constructs the object the safe way, namely by calling the constructor. The constructor requires an argument unfortunately, namely a service locator. How does the proxy know what this argument should be? Exactly, by storing the value parameter when Contants was created for the first time in a member field!

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java (hunk 6 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: see above

File: config/quality/findbugs/findbugs-filter.xml (hunk 12 above)
Comment (jverbraeken) [inline config/quality/findbugs/findbugs-filter.xml]: I don't think that you want to exclude classes from findbugs generally. We can always include those classes when that's really needed in the future

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java (hunk 20 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Yes, because Findbugs warns you that ContactListAdapter should be a private static inner class instead of an non-static outer class

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 22]
```diff
@@ -129,26 +128,28 @@ private void fillDatabaseWithDemoData() throws IOException {
...
-        final List<Contact> contacts = database.getAllContacts();
-        for (final Contact contact: contacts) {
-            database.deleteContact(contact.getName());
+        final Contact f = new Contact("Baas", new SimpleKey("Awesomeness", "nr23uinr3uin2o3uin23oi4un234ijn"));
+        if (!database.getAllContacts().isEmpty()) {
+            database.deleteContact(a.getName());
+            database.deleteContact(b.getName());
+            database.deleteContact(c.getName());
+            database.deleteContact(d.getName());
+            database.deleteContact(e.getName());
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: I would think that this gives errors. 'Baas' will be added every time and doesn't get deleted, and a,b,c,d and e get deleted even though they are not added

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 23]
```diff
@@ -129,26 +128,28 @@ private void fillDatabaseWithDemoData() throws IOException {
         final Contact c = new Contact("Joost", new SimpleKey("Webserver", "dnfh4nl4jknlkjnr4j34klnk3j4nl"));
         final Contact d = new Contact("Kilian", new SimpleKey("Webmail", "sdjnefiniwfnfejewjnwnkenfk32"));
         final Contact e = new Contact("Cornel", new SimpleKey("Awesomeness", "nr23uinr3uin2o3uin23oi4un234ijn"));
-
-        final List<Contact> contacts = database.getAllContacts();
-        for (final Contact contact: contacts) {
-            database.deleteContact(contact.getName());
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: I would put this back, explanation in the comment below

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 24]
```diff
@@ -129,26 +128,28 @@ private void fillDatabaseWithDemoData() throws IOException {
...
+            database.deleteContact(c.getName());
+            database.deleteContact(d.getName());
+            database.deleteContact(e.getName());
         }
-
-        database.addContact(a);
-        database.addContact(b);
-        database.addContact(c);
-        database.addContact(d);
-        database.addContact(e);
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: This belongs to the comment above

**Comment by jverbraeken** [comment]
@clenrock Fixed!

**Comment by TheBonheurs** [comment]
How can the coverage go down when @jverbraeken only increasde the amopunt of code tested?

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 25]
```diff
@@ -104,19 +110,13 @@ public void addContact(final Contact contact) throws IOException {
...
-        final int lengthBefore = contacts.size();
         for (final Contact contact : contacts) {
             if (contactName.equals(contact.getName())) {
                 contacts.remove(contact);
                 break;
             }
         }
 
-        // Throw if the contact to remove is not found
-        if (contacts.size() == lengthBefore) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Don't we want to know if the contact was not found and thus could not be deleted?

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java [hunk 26]
```diff
@@ -49,38 +53,6 @@ public RSAKey(final Map<String, String> map) throws IllegalArgumentException {
...
-    }
-
-    /**
-     * Get the modulus of an {@link RSAKey}.
-     *
-     * @param key An {@link RSAKey}.
-     * @return The value for the modulus attribute of the {@code key}.
-     * @throws IllegalArgumentException If the provided {@link IKey} is not a {@link RSAKey}.
-     */
-    static String getModulus(final IKey key) throws IllegalArgumentException {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java]: Why do you want to delete this?

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/tap/SingleTap.java [hunk 27]
```diff
@@ -22,4 +26,40 @@ public SingleTap() {
...
+     * Used for the Serialization Proxy Pattern.
+     * We suppress here the AccessorClassGeneration warning because the only alternative to this pattern -
+     * ordinary serialization - is far more dangerous
+     */
+    @SuppressWarnings("PMD.AccessorClassGeneration")
+    private static final class SerializationProxy implements Serializable {
+
+        private static final long serialVersionUID = 6955333968356740403L;
+
+        SerializationProxy() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/tap/SingleTap.java]: Missing JavaDoc (also applies to other instances of this similar internal classes)

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * This thread runs while listening for incoming connections. It behaves
+ * like a server-side client. It runs until a connection is accepted
+ * (or until cancelled).
+ */
+class AndroidAcceptThread extends Thread {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java]: Could you change this name so it is clear that it is about Bluetooth or make it more general?

File: app/src/main/java/com/nervousfish/nervousfish/service_locator/EntryActivity.java [hunk 29]
```diff
@@ -27,7 +27,7 @@ protected void onCreate(final Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
         final String androidFileDir = this.getFilesDir().getPath();
-        final IServiceLocator serviceLocator = new ServiceLocator(androidFileDir);
+        final ServiceLocator serviceLocator = new ServiceLocator(androidFileDir);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/EntryActivity.java]: Why not `IServiceLocator`  ?

File: config/quality/checkstyle/checkstyle.xml [hunk 30]
```diff
@@ -104,7 +104,9 @@
         <!-- See http://checkstyle.sourceforge.net/config_import.html -->
         <module name="AvoidStarImport" />
         <!-- We suppress this warning because the Java creators themselves disagree on this one, for example in the newest JUnit 5 -->
-        <module name="AvoidStaticImport" />
+        <module name="AvoidStaticImport">
```
Comment (ericcornelissen) [inline config/quality/checkstyle/checkstyle.xml]: Doesn't this only apply to the checkstyle ruleset for tests?

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java (hunk 26 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/RSAKey.java]: It isn't used

File: config/quality/checkstyle/checkstyle.xml (hunk 30 above)
Comment (jverbraeken) [inline config/quality/checkstyle/checkstyle.xml]: No, because I also use assertions in the code

Comment (ericcornelissen) [inline config/quality/checkstyle/checkstyle.xml]: Could you link to an example of that? 😄

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/tap/SingleTap.java (hunk 27 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/tap/SingleTap.java]: fixed

File: app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java [hunk 31]
```diff
@@ -280,13 +290,13 @@ public void onItemClick(final AdapterView<?> av, final View v, final int arg2, f
                 bluetoothHandler.connect(device);
 
                 // Create the result Intent and include the MAC address
-                final Intent intent = new Intent();
+                /*final Intent intent = new Intent();
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/BluetoothConnectionActivity.java]: Can you remove this and other commented code in this class? 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java [hunk 32]
```diff
@@ -128,29 +134,42 @@ private void fillDatabaseWithDemoData() throws IOException {
...
+    @Subscribe
+    public void onEvent(final ContactReceivedEvent event) {
+/*        try {
+            this.contacts = serviceLocator.getDatabase().getAllContacts();
+            final ListView lv = (ListView) findViewById(R.id.listView);
+            listviewAdapter.notifyDataSetChanged();
+            lv.setAdapter(new ContactListAdapter(this, this.contacts));
+        } catch (final IOException e) {
+            LOGGER.error("onResume in MainActivity threw an IOException");
+        }*/
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/MainActivity.java]: Can you either make this method useful or remove it? 😇

File: app/src/main/java/com/nervousfish/nervousfish/exceptions/DatabaseException.java [hunk 33]
```diff
@@ -0,0 +1,13 @@
+package com.nervousfish.nervousfish.exceptions;
+
+/**
+ * Thrown when there is a issue with the database
+ */
+public class DatabaseException extends RuntimeException {
+    /**
+     * {@inheritDoc}
+     */
+    public DatabaseException(final String msg) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/exceptions/DatabaseException.java]: Should `@override` when using `{@inheritDoc}`, right (applies to DeserializationException.java as well)?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java [hunk 34]
```diff
@@ -54,7 +56,7 @@ public void writeAllContacts() throws IOException {
      * @throws IOException When deserialization doesn't go well.
      */
     void writeContact(final Contact contact) throws IOException {
-        if (!checkExists(contact)) {
+        LOGGER.info("Begin writing contact :" + contact.getName());
             byte[] bytes = null;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java]: This indentation seems to be incorrect? Also, why are you removing this check?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java [hunk 35]
```diff
@@ -100,7 +101,8 @@ boolean checkExists(final Contact contact) throws IOException {
      * @param bytes byte array representing a contact
      * @return Whether or not the process finished successfully
      */
-    boolean saveContact(final byte[] bytes) {
+    Contact saveContact(final byte[] bytes) {
+        LOGGER.info("Saving these bytes :" + bytes);
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java]: ```LOGGER.info("Saving these bytes: " + bytes);```

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java [hunk 36]
```diff
@@ -110,30 +112,40 @@ boolean saveContact(final byte[] bytes) {
             contact = (Contact) ois.readObject();
         } catch (final ClassNotFoundException | IOException e) {
             LOGGER.error(" Couldn't start deserialization!");
-            return false;
+            e.printStackTrace();
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java]: Stacktrace should be printed using the logger (applies multiple times in this class)

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothHandler.java [hunk 37]
```diff
@@ -123,28 +127,11 @@ public void connect(final BluetoothDevice device) {
...
-                acceptThread.cancel();
-                acceptThread = null;
-            }
-
+    public synchronized void stop() {
+        /*LOGGER.info("Bluetooth service stopped");
             mBluetoothState = BluetoothState.STATE_NONE;
             getServiceLocator().postOnEventBus(new BluetoothConnectionLostEvent());
-        }
+        }*/
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothHandler.java]: Could you remove this commented code 😇

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java (hunk 34 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/APairingHandler.java]: I fixed the indentation! The check is certainly not needed, because the methods send a contact to its client. I changed the method to sendContact for clarity. The client will have to check for himself if he accepts or rejects the contact he receives.

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java [hunk 38]
```diff
@@ -20,22 +17,23 @@
...
     /**
      * Constructs a new AndroidAcceptThread that runs while listening for incoming connections. It behaves
      * like a server-side client. It runs until a connection is accepted
      * (or until cancelled).
      *
+     * @param bluetoothHandler The {@link AndroidBluetoothHandler} owning this thread
      * @param serviceLocator The service locator to be used
      */
-    AndroidAcceptThread(final IServiceLocator serviceLocator) {
+    AndroidAcceptThread(final AndroidBluetoothHandler bluetoothHandler, final IServiceLocator serviceLocator) {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java]: Why not get the `bluetoothHandler` from the `serviceLocator`?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java [hunk 39]
```diff
@@ -67,11 +66,12 @@ public void run() {
...
                 socket = serverSocket.accept();
             } catch (final IOException e) {
-                LOGGER.error("Listening on the socket failed");
+                LOGGER.error("Listening on the socket failed", e);
                 return;
             }
 
-            serviceLocator.postOnEventBus(new BluetoothAlmostConnectedEvent(socket));
+            this.bluetoothHandler.notifyAlmostConnected(socket);
+            //serviceLocator.postOnEventBus(new BluetoothAlmostConnectedEvent(socket));
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java]: Can you remove commented code 😇

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothHandler.java [hunk 40]
```diff
@@ -16,28 +17,29 @@
...
 import java.util.Arrays;
 
-import static org.junit.Assert.assertNotNull;
-
 /**
  * This Bluetooth service class establishes and manages a bluetooth connection.
  */
 
-@SuppressWarnings({"checkstyle:classdataabstractioncoupling", "PMD.NullAssignment", "PMD.TooManyMethods"})
+@SuppressWarnings({"checkstyle:classdataabstractioncoupling", "PMD.NullAssignment", "PMD.TooManyMethods", "LawOfDemeter"})
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidBluetoothHandler.java]: Add a comment for the new suppression please 😄

File: build.gradle [hunk 41]
```diff
@@ -8,6 +8,7 @@ buildscript {
         classpath 'com.android.tools.build:gradle:2.3.2'
         classpath 'org.kt3k.gradle.plugin:coveralls-gradle-plugin:2.8.1'
         classpath 'net.saliman:gradle-cobertura-plugin:2.4.0'
+        classpath 'de.aaschmid:gradle-cpd-plugin:1.0'
```
Comment (ericcornelissen) [inline build.gradle]: Can you give me a link to the file where this is used? 😇

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java (hunk 38 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/AndroidAcceptThread.java]: Because I didn't get the Eventbus working, as you saw yesterday. Now I'm using a callback, but I want that only the helper classes like AndroidAcceptThread can call these. Therefore I cannot implement them in the interface.

File: .idea/inspectionProfiles/Project_Default.xml [hunk 42]
```diff
@@ -974,7 +974,6 @@
       <option name="ignoreParenthesesOnLambdaParameter" value="false" />
     </inspection_tool>
     <inspection_tool class="UnnecessaryQualifierForThis" enabled="true" level="WARNING" enabled_by_default="true" />
-    <inspection_tool class="UnnecessarySuperConstructor" enabled="true" level="WARNING" enabled_by_default="true" />
```
Comment (ericcornelissen) [inline .idea/inspectionProfiles/Project_Default.xml]: Why did you remove this line?

File: .idea/inspectionProfiles/Project_Default.xml [hunk 43]
```diff
@@ -1,43 +1,45 @@
 <component name="InspectionProjectProfileManager">
```
Comment (ericcornelissen) [inline .idea/inspectionProfiles/Project_Default.xml]: Can this file be removed form this PR please 😇

**Comment by jverbraeken** [comment]
@ericcornelissen Done!
